docs/dificil/small_to_large.md

```diff
@@ -1,7 +1,225 @@
+
+<!-- Demonstratie scurta -->
+<!-- Ne putem baza pe același principiu ca la `DSU`, anume că dacă la un moment dat avem un element $x$ într-un set de mărime $s$, în cel mai rău caz, acesta va fi reunit cu un set tot de mărime $s$. Astfel, următorul set în care se va afla elementul va avea cel puțin $2s$ elemente, sau altfel spus, orice element va fi inserat în maxim $log(n)$ seturi pe parcursul algoritmului. Așadar, vom avea complexitate $O(nlog^2(n))$, deoarece și inserările iau $O(log(n))$. -->
+
+Cu această schimbare, codul va arăta așa:
+
+```c++
+#include <iostream>
+#include <set>
+#include <vector>
+
+const int NMAX = 200005;
+int rez[NMAX];
+std::vector <int> G[NMAX];
+//tinem minte cate un set pentru fiecare nod
+std::set <int> s[NMAX];
+
+void dfs(int nod, int t) {
+    for (auto x : G[nod]) {
+        if (x != t) {
+            dfs(x, nod);
+            //daca numarul de elemente din setul fiului este mai mare decat cel din setul nodului, le interschimbam
+            if(s[x].size() > s[nod].size())
+                s[nod].swap(s[x]);
+            //parcurgem setul fiecarui fiu si adaugam elementele in setul nodului
+            for (auto x2 : s[x])
+                s[nod].insert(x2);
+        }
+    }
+    //obtinem rezultatul folosind metoda .size()
+    rez[nod] = s[nod].size();
+}
+
+int main() {
+    int n, i, u, v, c;
+    std::cin >> n;
+    for (i = 1; i <= n; i++) {
+        std::cin >> c;
+        //valoarea nodului o putem adauga direct in setul acestuia
+        s[i].insert(c);
+    }
+    for (i = 1; i < n; i++) {
+        std::cin >> u >> v;
+        G[u].push_back(v);
+        G[v].push_back(u);
+    }
+    dfs(1, 0);
+    for (i = 1; i <= n; i++)
+        std::cout << rez[i] << " ";
+    return 0;
+}
+
+```
+
+Această soluție se încadrează în limitele problemei și va lua punctaj maxim.
+
+
+## Small to large folosind `__gnu_pbds`
+
+Această structură de date oferă aceleași funcții ca structura `set` din STL, însă mai oferă încă 2 funcții care ne pot fi de ajutor în anumite probleme:
+
+1. `find_by_order(x)`, care returnează un iterator către a x-a cheie ca ordine, în timp $O(log(n))$.
+2. `order_of_key(x)`, care returnează numărul de elemente **strict** mai mici ca x, tot în timp $O(log(n))$.
+
+Pentru a putea folosi această structură, va trebui să adăugăm la programul nostru următoarele linii:
+```c++
+#include <ext/pb_ds/assoc_container.hpp>
+#include <ext/pb_ds/tree_policy.hpp>
+using namespace __gnu_pbds;
+#define ordered_set tree<int, null_type, std::less_equal<int>, rb_tree_tag, tree_order_statistics_node_update>
```

Comment:
Fix linia asta... ugh. Dacă tot vrei să mergi pe calea asta, folosește măcar un typedef sau (de preferat) un using:

```cpp
using ordered_set = tree<int, null_type, std::less_equal<int>, rb_tree_tag, tree_order_statistics_node_update>;
```

Așa devine un tip real și nu un find and replace glorificat.